src/server/accept.rs

```diff
@@ -89,83 +74,111 @@ where
-                    Some(false) => 0.into(),
-                    Some(true) | None => 1.into(),
+            });
+
+            tokio::select! {
+                // A connection was accepted or IDLE_TIMEOUT expired.
+                result = accept => match result {
+                    Ok(accepted) => accepted,
+                    Err(error) => {
+                        eprintln!("error(listener): {}", error);
+                        drop(permit);
+                        continue;
+                    }
+                },
+                // A shutdown signal was received.
+                _ = shutdown_rx.changed() => {
+                    drop(permit);
+                    break 'accept match *shutdown_rx.borrow_and_update() {
+                        Some(false) => 0.into(),
+                        Some(true) | None => 1.into(),
+                    }
                 }
             }
         };
 
-        // Spawn a task to serve the connection.
-        connections.spawn({
-            // Clone acceptor so negotiation can happen in the connection task.
+        connections.push_back({
             let mut acceptor = acceptor.clone();
-
-            // Clone the watch receiver so we can shutdown the connection if a
-            // ctrl+c signal is sent to the process.
             let mut shutdown_rx = shutdown_rx.clone();
 
-            // Create an AppService to serve the connection.
-            let service = AppService::new(Arc::clone(&app), max_body_size);
-
-            async move {
-                let result = match acceptor.accept(stream).await {
-                    Err(error) => Err(error.into()),
-                    Ok(accepted) => {
-                        // Create a new HTTP/2 connection.
-                        #[cfg(feature = "http2")]
-                        let connection = conn::http2::Builder::new(TokioExecutor::new())
-                            .timer(TokioTimer::new())
-                            .serve_connection(IoStream::new(accepted), service);
-
-                        // Create a new HTTP/1.1 connection.
-                        #[cfg(all(feature = "http1", not(feature = "http2")))]
-                        let connection = conn::http1::Builder::new()
-                            .timer(TokioTimer::new())
-                            .serve_connection(IoStream::new(accepted), service)
-                            .with_upgrades();
-
-                        tokio::pin!(connection);
-
-                        // Serve the connection.
-                        tokio::select! {
-                            result = &mut connection => result.map_err(|e| e.into()),
-                            _ = shutdown_rx.changed() => {
-                                connection.as_mut().graceful_shutdown();
-                                connection.await.map_err(|e| e.into())
-                            }
-                        }
+            // TODO: Attempt to get the task size in the tls example <= 1024.
```

Comment:
This is probably something we want to be cognizant of.

It's probably safe to consider anything under `2048` a reasonable task size for an HTTPS server. If my memory serves me correctly, that is the norm.

It's also somewhat likely that a smaller task size in general is more desirable than avoiding an allocation in the accept loop. Especially if our intention is to yield to give the task we attempted to join in the select loop time to finish.

I'm unlikely to make a decision here until I perform some serious benchmarks and profiling on Linux. I've been developing exclusively on macOS as I do a lot of React Native work professionally.

Reply:
FWIW, I have a hunch that app servers should always prefer the latter and the numbers I've been coming up with are default values that I saw countless times in apache / nginx config files.